### PR TITLE
ARROW-13100: [MATLAB] Integrate GoogleTest with MATLAB Interface C++ Code

### PR DESCRIPTION
**Pull request by kevingurney**
## Overview

This Pull Request:

1. **Adds additional options for integrating the Arrow C++ libraries with the MATLAB interface.**

    1.1. Users can supply a custom `ARROW_HOME` value to make use of their own, pre-built `libarrow` library.

    1.2. If no information is provided, then CMake will attempt to find a system-installed Arrow distribution (e.g. installed via a package manager).

    1.3. If no system Arrow libraries are found by `find_package`, and no additional information is provided, then the Arrow C++ libraries will automatically be built from source using `ExternalProject_Add`.

2. **Adds support for GoogleTest to the CMake build system for the MATLAB interface.**

    2.1.  A user can supply a custom [`GTEST_ROOT`](https://cmake.org/cmake/help/latest/module/FindGTest.html) value to CMake in order to make use of pre-built GoogleTest binaries.

    2.2. If no information is provided, then CMake will attempt to find a system-installed GoogleTest distribution (e.g. installed via a package manager).

    2.3. If no system GoogleTest libraries are found by `find_package`, and no additional information is provided, then the CMake build system will use the bundled GoogleTest binaries that are built as part of the Arrow C++ libraries ( `ARROW_BUILD_TESTS` is set to `ON` under the hood).

3. **Adds a `MATLAB_BUILD_TESTS` option to the CMake build system for the MATLAB interface.**

    3.1. This allows users to enable building of the  C++ tests for the MATLAB interface with `cmake .. -D MATLAB_BUILD_TESTS=ON`.

4. **Allows the MATLAB interface and the C++ tests to be built and run using a consistent set of commands on Windows, macOS, and Linux.**

```bash
$ cmake -S . -B build -D MATLAB_BUILD_TESTS=ON
$ cmake --build build --config Release
$ ctest --test-dir build
```

## Implementation

This implementation uses [`ExternalProject_Add`](https://cmake.org/cmake/help/latest/module/ExternalProject.html#command:externalproject_add) to automatically build the Arrow C++ libraries and bundled GoogleTest binaries from source.

We took heavy inspiration from the CMake patterns used to create imported library targets in [`ThirdpartyToolchain.cmake`](https://github.com/apache/arrow/blob/master/cpp/cmake_modules/ThirdpartyToolchain.cmake).

We followed the [same approach used in the CMake build system for the Arrow C++ libraries](https://github.com/apache/arrow/blob/e2238582e2a2bf20a68a967145fe1a7b2337a997/cpp/cmake_modules/ThirdpartyToolchain.cmake#L1701) and added an automatic step which copies the `gtest.dll` and `gtest_main.dll` runtime libraries to the same folder as the C++ tests for the MATLAB interface, so that they can be found by the [Windows load-time dynamic linker](https://docs.microsoft.com/en-us/windows/win32/dlls/load-time-dynamic-linking).
 
## Testing

1. These changes were qualified against the following platform/compiler configurations:
 1.1. Windows 10 with Visual Studio 2019
 1.2. macOS Big Sur (11.2.3) with GNU Make 3.81
 1.3. Debian 10 with GNU Make GNU 4.2.1

2. We ran the CMake build with a custom `GTEST_ROOT` value that points to GoogleTest binaries that were built via the CMake `googletest_ep` target from the Arrow C++ libraries. In other words, we built GoogleTest with `cmake -DARROW_BUILD_TESTS=ON  ..; make googletest_ep`.
3. We ran the CMake build without specifying a `GTEST_ROOT` value and the GoogleTest binaries that are bundled with the Arrow C++ libraries (when built with `-D ARROW_BUILD_TESTS=ON`) were automatically used.
4. We ran the CMake build with `MATLAB_BUILD_TESTS=ON` and verified that an Arrow installation found by `find_package` (e.g. when an explicit `ARROW_HOME` is specified) is used, instead of the Arrow library built from source, using `ldd` on Linux, `otool -l` and `otool -L` on macOS, and the [`/VERBOSE:LIB`](https://docs.microsoft.com/en-us/cpp/build/reference/verbose-print-progress-messages?view=msvc-160#remarks) linker flag on Windows. We also verified that deleting the `arrow.lib` import library from a custom `ARROW_HOME` resulted in a linker failure on Windows.
5. We built and ran the tests on Windows, macOS, and Linux using the following commands:

```
$ cmake -S . -B build -D MATLAB_BUILD_TESTS=ON
$ cmake --build build --config Release
$ ctest --test-dir build
```

## Future Directions

1. As follow up work to this PR, a major priority is to integrate the MATLAB interface with the Arrow CI ecosystem, so that both C++ and MATLAB tests will be automatically run with every pull request.
2. Pending acceptance of this pull request, the MATLAB interface will have C++ testing infrastructure in place. This will allow us to shift focus towards incremental delivery of C++ feature code, along with associated tests.
3. We plan to update the MATLAB interface documentation to clearly explain how to build and run the C++ tests with and without a custom `GTEST_ROOT` value. This work is captured in [ARROW-13204](https://issues.apache.org/jira/browse/ARROW-13204).
4. The current implementation requires the C++ tests for the Arrow C++ libraries to be built in order to use the bundled GoogleTest binaries. Building the Arrow C++ tests takes a nontrivial amount of time and isn't strictly necessary to validate the correctness of the MATLAB interface. In the future, we may want to consider fetching and building the GoogleTest binaries from GitHub separately. One approach to achieve this would be to factor out the [`build_gtest` macro](https://github.com/apache/arrow/blob/e2238582e2a2bf20a68a967145fe1a7b2337a997/cpp/cmake_modules/ThirdpartyToolchain.cmake#L1602) from `ThirdpartyToolchain.cmake` and make it reusable by the MATLAB interface.
5. Automatically copying `gtest.dll` and `gtest_main.dll` to the same folder as the C++ tests during the build process helps new developers get started with less friction on Windows. Copying `arrow.dll` to the same folder as the MEX files may similarly be useful. This would prevent users from having to manually add `arrow.dll` to the Windows `%PATH%`. This would also require us to keep all the MEX files in the build folder. There would be some trade-offs with this approach concerning MEX file organization and discoverability on the MATLAB Path, so this would require a bit more thought.
6. This PR doesn't support debug builds of GoogleTest when building the bundled GoogleTest binaries. We may want to support this in the future.
7. When specifying a custom `GTEST_ROOT`, `gtest.dll` and `gtest_main.dll` won't be automatically copied to the location of the C++ tests. Therefore, they won't be discoverable by the Windows load-time dynamic linker by default. It might be nice to support automatic copying for this use case in the future.
 
## Notes

1. Thank you to @lafiona for helping me with this pull request!

### CODE REVIEW

**Comment by kevingurney** [comment]
My apologies! It looks like I forgot to add the Apache license header to `matlab/src/placeholder_test.cc`. I will make that change now.

**Comment by kevingurney** [comment]
The most recent CI failures seem to be related to `cmake-format` checks failing on `arrow/cpp/cmake_modules/ThirdpartyToolchain.cmake`:

```bash
ERROR __main__.py:618: Check failed: /arrow/cpp/cmake_modules/ThirdpartyToolchain.cmake
```

It didn't seem like this file was supposed to be formatted using `cmake-format` since a large portion of the lines that were unmodified by this PR did not match the expected style when we originally tried running the CMake linting checks using `archery`.

Given this fact, I'm assuming its safe to ignore this CI failure, but let me know if I should also format `ThirdpartyToolchain.cmake` using `cmake-format`.

File: matlab/CMakeLists.txt [hunk 1]
```diff
@@ -17,6 +17,23 @@
...
+  # The ExternalProject module is required for downloading and building external dependencies.
+  include(ExternalProject)
+  # The ThirdpartyToolchain module automatically builds xsimd by default,
+  # so we explicitly disable building of xsimd here.
+  set(ARROW_SIMD_LEVEL "NONE")
+  set(ARROW_RUNTIME_SIMD_LEVEL "NONE")
+  # The ThirdpartyToolchain module must know the location of the Arrow C++ libraries source code.
+  set(arrow_SOURCE_DIR ${CMAKE_SOURCE_DIR}/../cpp)
+  # The ThirdpartyToolchain module uses ExternalProject to download GoogleTest.
+  include(ThirdpartyToolchain)
```
Comment (kou) [inline matlab/CMakeLists.txt:32]: `ThirdpartyToolchain isn't expected to be used by `include`.
Could you use `find_package(GTest REQUIRED CONFIG)` instead?

File: matlab/CMakeLists.txt [hunk 2]
```diff
@@ -72,3 +89,28 @@ else()
...
+
+# ######################
+# GoogleTest Integration
+# ######################
+# If the user has specified a GTEST_ROOT value, use their pre-built GoogleTest binaries.
+# Otherwise, download and build GoogleTest automatically.
+if(GTEST_ROOT)
+  find_package(GTest REQUIRED)
+else()
+  download_and_build_google_test()
```
Comment (kou) [inline matlab/CMakeLists.txt:101]: Do we need this functionality for MATLAB module?
How about installing GoogleTest via package manager on system such as `apt`, `dnf`, `brew` and so on?

File: matlab/CMakeLists.txt (hunk 1 above)
Comment (kevingurney) [reply matlab/CMakeLists.txt:32]: Thanks for reviewing these changes, @kou!

Could you elaborate on why `ThirdpartyToolchain` isn't expected to be `include`d in other CMake files?

We intentionally wanted to reuse the robust infrastructure that the community took time to develop for finding 3rd party libraries in a flexible way, rather than reinventing the wheel here.

Comment (kou) [reply matlab/CMakeLists.txt:32]: `ThirdpartyToolchain` has many Apache Arrow C++ specific configurations like `configure_file()` you changed.
So you need to set some `ARROW_*` variables to `include` `ThirdpartyToolchain`.

If we want to reuse `ThirdpartyToolchain` in the MATLAB library, we should make `ThirdpartyToolchain` reusable and reuse rather than `set` some variables and `include`.

File: matlab/CMakeLists.txt (hunk 2 above)
Comment (kou) [reply matlab/CMakeLists.txt:101]: OK.
How about referring GoogleTest bundled in Apache Arrow C++ instead of adding this functionality to the MATLAB library?
If we build bundled GoogleTest in Apache Arrow C++, it's installed into `${ARROW_BUILD_DIR}/googletest_ep-prefix/{include,lib}/`. We can find it with `find_package(GTest REQUIRED CONFIG)` and `-DGTest_ROOT=${ARROW_BUILD_DIR}/googletest_ep-prefix`.
And `find_package(GTest REQUIED COFNIG)` without `-DGTest_ROOT` will find GoogleTest installed into system by package manager.

Comment (kevingurney) [reply matlab/CMakeLists.txt:101]: Setting `GTEST_ROOT` to `${ARROW_BUILD_DIR}/googletest_ep-prefix` inside the `CMakeLists.txt` for MATLAB seems like a reasonable approach.

If we could automatically trigger a build of the Arrow C++ libraries with `ARROW_BUILD_TESTS=ON` (in order to get GoogleTest to be built in `${ARROW_BUILD_DIR}/googletest_ep-prefix`) from the MATLAB `CMakeLists.txt` file, that would be great. But, I'm not sure what the best approach is to accomplish this.

It seems like it would be error prone to expect users to manually build the Arrow C++ libraries first due to the variation in names they might use for their CMake build folder, as well as the different flags they might pass to the build (e.g. they might forget to include `-DARROW_BUILD_TESTS=on`).

File: matlab/CMakeLists.txt (hunk 1 above)
Comment (kevingurney) [reply matlab/CMakeLists.txt:32]: See my comment below.

It sounds like depending on `ThirdpartyToolchain` may not be desirable without first refactoring it to make it more reusable. Given this fact, I'm fine with instead pursuing the approach of trying to automatically build the Arrow C++ libraries first, and then depending on the bundled GoogleTest binaries that are built when `ARROW_BUILD_TESTS` is set to `on`.

File: matlab/CMakeLists.txt (hunk 2 above)
Comment (kevingurney) [reply matlab/CMakeLists.txt:101]: We tried invoking an automatic build of the Arrow C++ libraries from `matlab/CMakeLists.txt` using:

```cmake
ExternalProject_Add(ArrowCpp SOURCE_DIR "${CMAKE_SOURCE_DIR}/../cpp" CMAKE_ARGS -DCMAKE_INSTALL_PREFIX=. -DARROW_BUILD_TESTS=ON)
```

But, unfortunately, this doesn't seem to be exporting targets for `arrow_shared` or `GTest::GTest`. Admittedly, I am not particularly familiar with `ExternalProject_Add`.

Is this the approach you were suggesting? Or, were you thinking that users would just manually build the Arrow C++ libraries first, in order to get access to the bundled GoogleTest binares (i.e. similar to https://github.com/apache/arrow/blob/master/docs/source/developers/python.rst#build-and-test)?

Comment (kou) [reply matlab/CMakeLists.txt:101]: Can we clarify target people of the MATLAB library's test?

1. Users: People who use the MATLAB library.
2. Developers: People who develop the MATLAB library.
3. Both of 1. and 2.

My thought is 2.

For 2., custom Apache Arrow C++ build is useful. For example, developers may want to build with `-DCMAKE_BUILD_TYPE=debug` for debugging and with `-DCMAKE_BUILD_TYPE=release` for benchmarking.

So, it may be reasonable that developers build Apache Arrow C++ manually.

For 1., users may not want to enable tests by default. Because they just want to use the MATLAB library.

If you still want to implement automatic Apache Arrow C++ build in CMake, I'll take a look.

Comment (kevingurney) [reply matlab/CMakeLists.txt:101]: Thanks for sharing your thoughts, @kou!

-------------

> Can we clarify target people of the MATLAB library's test?

I agree with you that _Developers_ are the primary target audience of the C++ tests for the MATLAB interface.

> For 2., custom Apache Arrow C++ build is useful. For example, developers may want to build with -DCMAKE_BUILD_TYPE=debug for debugging and with -DCMAKE_BUILD_TYPE=release for benchmarking

You make an excellent point about providing developers with the flexibility to build their own custom `libarrow` for debugging and benchmarking purposes! This is definitely something that we would like to support. If we use `find_package`, as you previously suggested, this should support custom `libarrow` builds via `ARROW_HOME`, as well as package managers.

> For 1., users may not want to enable tests by default. Because they just want to use the MATLAB library.

Good point! I think it would make sense to introduce a flag like `MATLAB_BUILD_TESTS`, with a default value of `OFF`, so that users can toggle building of the C++ tests for the MATLAB interface.

-------------

Based on your input, we went back to the drawing board and learned more about `ExternalProject_Add` and `IMPORTED` targets. By studying the source code for `build_dependency` in `ThirdpartyToolchain` more closely, I now understand how `ExternalProject_Add` can be used to trigger an automatic build of the Apache Arrow C++ libraries (and bundled GoogleTest).

I put together a prototype which automatically builds Arrow C++ and bundled GoogleTest on Linux (only if the user has specified `-DMATLAB_BUILD_TESTS=ON`).

**You can find the prototype here:**

https://github.com/mathworks/arrow/commit/42609d9671501c8fe1e8c613337eedb207d32e91

I tested this manually under the following conditions:

1. GoogleTest installed via the `apt` package manager to `/usr/lib/...` on Debian 10
2. A custom `GTEST_ROOT` and a custom `ARROW_HOME` value are specified
3. `find_package` fails to find a valid Arrow or GTest installation via 1. or 2., and an automatic C++ source build is triggered

The code could probably benefit from some additional refactoring, and I'm not sure whether "Modern CMake Best Practices" would deem it acceptable to depend on the bundled GoogleTest binaries in this way, but it does seem to be working as expected.

Let me know what you think. 

Thanks!

Comment (kou) [reply matlab/CMakeLists.txt:101]: Could you merge the commit to this branch?
Then I can comment on the implementation in this pull request.

Comment (kevingurney) [reply matlab/CMakeLists.txt]: @kou - my apologies for the delay! I was away for a few days.

I've pushed the changes from mathworks@42609d9 to this branch, so that you can comment on the implementation.

Thank you!

File: matlab/CMakeLists.txt [hunk 3]
```diff
@@ -17,6 +17,84 @@
 
 cmake_minimum_required(VERSION 3.20)
 
+# Build the Arrow C++ libraries WITHOUT bundled GoogleTest binaries.
+function(BUILD_ARROW)
```
Comment (kou) [inline matlab/CMakeLists.txt:21]: Lowercase is better because we use lowercase for function call:

```suggestion
function(build_arrow)
```

Comment (kou) [reply matlab/CMakeLists.txt:21]: Can we unify `build_arrow` and `build_arrow_and_gtest` by using function arguments?
We can use `cmake_parse_arguments()` for processing arguments: https://cmake.org/cmake/help/latest/command/cmake_parse_arguments.html
See also `resolve_dependency` macro that uses `cmake_parse_arguments()`: https://github.com/apache/arrow/blob/master/cpp/cmake_modules/ThirdpartyToolchain.cmake#L200-L204

File: matlab/CMakeLists.txt [hunk 4]
```diff
@@ -17,6 +17,84 @@
...
+  include(ExternalProject)
+
+  set(ARROW_PREFIX "${CMAKE_CURRENT_BINARY_DIR}/arrow_ep-prefix")
+  set(ARROW_INCLUDE_DIR "${ARROW_PREFIX}/include")
+  set(ARROW_LIBRARY_DIR "${ARROW_PREFIX}/lib")
+  set(ARROW_SHARED_LIB "${ARROW_LIBRARY_DIR}/libarrow.so")
```
Comment (kou) [inline matlab/CMakeLists.txt:27]:
```suggestion
  set(ARROW_SHARED_LIB "${ARROW_LIBRARY_DIR}/${CMAKE_SHARED_LIBRARY_PREFIX}arrow${CMAKE_SHARED_LIBRARY_SUFFIX}")
```

File: matlab/CMakeLists.txt [hunk 5]
```diff
@@ -17,6 +17,84 @@
...
+  set(ARROW_BINARY_DIR "${CMAKE_CURRENT_BINARY_DIR}/arrow_ep_build")
+
+  externalproject_add(arrow_ep
+                      SOURCE_DIR "${CMAKE_SOURCE_DIR}/../cpp"
+                      BINARY_DIR "${ARROW_BINARY_DIR}"
+                      CMAKE_ARGS -DCMAKE_INSTALL_PREFIX=${ARROW_PREFIX}
```
Comment (kou) [inline matlab/CMakeLists.txt:33]: We need to specify `CMAKE_INSTALL_LIBDIR=lib` explicitly because some platforms such as CentOS uses `lib64` for the default `CMAKE_INSTALL_LIBDIR`:

```suggestion
                      CMAKE_ARGS -DCMAKE_INSTALL_PREFIX=${ARROW_PREFIX} -DCMAKE_INSTALL_LIBDIR=lib
```

File: matlab/CMakeLists.txt [hunk 6]
```diff
@@ -17,6 +17,84 @@
...
 
+# Build the Arrow C++ libraries WITHOUT bundled GoogleTest binaries.
+function(BUILD_ARROW)
+  include(ExternalProject)
+
+  set(ARROW_PREFIX "${CMAKE_CURRENT_BINARY_DIR}/arrow_ep-prefix")
+  set(ARROW_INCLUDE_DIR "${ARROW_PREFIX}/include")
+  set(ARROW_LIBRARY_DIR "${ARROW_PREFIX}/lib")
+  set(ARROW_SHARED_LIB "${ARROW_LIBRARY_DIR}/libarrow.so")
+  set(ARROW_BINARY_DIR "${CMAKE_CURRENT_BINARY_DIR}/arrow_ep_build")
```
Comment (kou) [inline matlab/CMakeLists.txt:28]:
```suggestion
  set(ARROW_BINARY_DIR "${CMAKE_CURRENT_BINARY_DIR}/arrow_ep-build")
```

File: matlab/CMakeLists.txt [hunk 7]
```diff
@@ -32,8 +110,29 @@ endif()
 
 set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${CMAKE_SOURCE_DIR}/cmake_modules)
 
-# Arrow is Required
-find_package(Arrow REQUIRED)
+# Only build the MATLAB interface C++ tests if MATLAB_BUILD_TESTS=ON.
+if(MATLAB_BUILD_TESTS)
+  # Should support custom GTEST_ROOT as well as package managers.
```
Comment (kou) [inline matlab/CMakeLists.txt:115]: `-DGTest_ROOT` should work with only `find_package(GTest)`. We don't need additional code for it: https://cmake.org/cmake/help/latest/command/find_package.html#search-procedure

File: matlab/CMakeLists.txt [hunk 8]
```diff
@@ -32,8 +110,29 @@ endif()
...
+  else()
+    # Should support custom ARROW_HOME as well as package managers.
+    find_package(Arrow)
+    if(NOT Arrow_FOUND)
+      # Trigger an automatic build of the Arrow C++ libraries.
+      build_arrow()
+    endif()
+  endif()
+else()
+  # Should support custom ARROW_HOME as well as package managers.
```
Comment (kou) [inline matlab/CMakeLists.txt:129]: `ARROW_HOME` is processed by `FindArrow.cmake`. So we don't need additional code here.

File: matlab/CMakeLists.txt [hunk 9]
```diff
@@ -32,8 +110,29 @@ endif()
 
 set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${CMAKE_SOURCE_DIR}/cmake_modules)
 
-# Arrow is Required
-find_package(Arrow REQUIRED)
+# Only build the MATLAB interface C++ tests if MATLAB_BUILD_TESTS=ON.
+if(MATLAB_BUILD_TESTS)
```
Comment (kou) [inline matlab/CMakeLists.txt:114]: Could you add `option(MATLAB_BUILD_TESTS ...)` somewhere?
https://cmake.org/cmake/help/latest/command/option.html

File: matlab/CMakeLists.txt (hunk 7 above)
Comment (kevingurney) [reply matlab/CMakeLists.txt:115]: My apologies - I think the wording of my comment on line 115 may have lead to some confusion. What I meant by this comment was "`find_package(GTest) should already automatically support GTEST_ROOT and package managers`, not that we "should add more code to enable this".

I'll reword this comment to be less ambiguous.

File: matlab/CMakeLists.txt (hunk 8 above)
Comment (kevingurney) [reply matlab/CMakeLists.txt:129]: See my comment above. The comment wording here is misleading. I'll reword this to be less ambigious.

File: matlab/CMakeLists.txt (hunk 9 above)
Comment (kevingurney) [reply matlab/CMakeLists.txt:114]: Good point! Thanks for the suggestion. I'll make this chance.

File: matlab/CMakeLists.txt (hunk 5 above)
Comment (kevingurney) [reply matlab/CMakeLists.txt:33]: Thanks for pointing this out! I didn't realize that some platforms will install shared libraries to directories other than `lib` by default.

File: matlab/CMakeLists.txt (hunk 4 above)
Comment (kevingurney) [reply matlab/CMakeLists.txt:27]: Thanks for this suggestion. This will help make the code more generic across platforms.

File: matlab/CMakeLists.txt (hunk 7 above)
Comment (kevingurney) [reply matlab/CMakeLists.txt]: Fixed.

File: matlab/CMakeLists.txt (hunk 8 above)
Comment (kevingurney) [reply matlab/CMakeLists.txt:131]: Fixed.

File: matlab/CMakeLists.txt (hunk 9 above)
Comment (kevingurney) [reply matlab/CMakeLists.txt:116]: Fixed.

File: matlab/CMakeLists.txt (hunk 3 above)
Comment (kevingurney) [reply matlab/CMakeLists.txt]: Thanks - this is a really good idea!

We factored out the GoogleTest-specific logic into a macro called `build_gtest` for code cleanliness and added a `BUILD_GTEST` argument to `build_arrow` in order to enable us to remove the `build_arrow_and_gtest` function. `build_arrow` now just calls `build_gtest` when `BUILD_GTEST` is supplied.

**Comment by kevingurney** [comment]
Thanks, @kou for all of your really helpful feedback! We've addressed all of your comments.

We will now focus on adding the additional logic to make this work on Windows and macOS. This hopefully shouldn't take too much effort.

File: matlab/CMakeLists.txt [hunk 10]
```diff
@@ -17,13 +17,95 @@
...
+
+  add_dependencies(arrow_shared arrow_ep)
+endfunction()
+
+# Build the GoogleTest binaries that are bundled with the Arrow C++ libraries.
+macro(build_gtest)
+  set(ARROW_GTEST_PREFIX "${ARROW_BINARY_DIR}/googletest_ep-prefix")
+  set(ARROW_GTEST_INCLUDE_DIR "${ARROW_GTEST_PREFIX}/include")
+  set(ARROW_GTEST_LIBRARY_DIR "${ARROW_GTEST_PREFIX}/lib")
+  set(ARROW_GTEST_SHARED_LIB "${ARROW_GTEST_LIBRARY_DIR}/libgtest.so")
```
Comment (kou) [inline matlab/CMakeLists.txt:71]:
```suggestion
  set(ARROW_GTEST_SHARED_LIB "${ARROW_GTEST_LIBRARY_DIR}/${CMAKE_SHARED_LIBRARY_PREFIX}gtest${CMAKE_SHARED_LIBRARY_SUFFIX}")
```

File: matlab/CMakeLists.txt [hunk 11]
```diff
@@ -17,13 +17,95 @@
...
+
+  set(ARROW_GTEST_MAIN_PREFIX "${ARROW_BINARY_DIR}/googletest_ep-prefix")
+  set(ARROW_GTEST_MAIN_INCLUDE_DIR "${ARROW_GTEST_MAIN_PREFIX}/include")
+  set(ARROW_GTEST_MAIN_LIBRARY_DIR "${ARROW_GTEST_MAIN_PREFIX}/lib")
+  set(ARROW_GTEST_MAIN_SHARED_LIB "${ARROW_GTEST_MAIN_LIBRARY_DIR}/libgtest_main.so")
```
Comment (kou) [inline matlab/CMakeLists.txt:76]:
```suggestion
  set(ARROW_GTEST_MAIN_SHARED_LIB "${ARROW_GTEST_MAIN_LIBRARY_DIR}/${CMAKE_SHARED_LIBRARY_PREFIX}gtest_main${CMAKE_SHARED_LIBRARY_SUFFIX}")
```

File: matlab/CMakeLists.txt (hunk 7 above)
Comment (kou) [reply matlab/CMakeLists.txt]: Ah, sorry for my misunderstanding.

**Review by kou — COMMENTED** [review]
> We will now focus on adding the additional logic to make this work on Windows and macOS. This hopefully shouldn't take too much effort.

OK. Can we merge this once this works on Linux, Windows and macOS?

File: matlab/CMakeLists.txt (hunk 7 above)
Comment (kevingurney) [reply matlab/CMakeLists.txt]: No need to apologize! The original wording was confusing.

**Comment by kevingurney** [comment]
Yes - we are actively working on adding support for Windows and macOS, now. We'll add the changes to this branch.

Thanks!

**Comment by kevingurney** [comment]
@kou we just pushed updates to `matlab/CMakeLists.txt` which enable building of the Arrow C++ libraries and bundled GoogleTest binaries automatically from source on macOS and Windows.

We are now able to build and run the MATLAB C++ tests consistently, using the same set of commands, on Windows, Mac, and Linux:

```bash
$ cmake -S . -B build -D MATLAB_BUILD_TESTS=ON
$ cmake --build build --config Release
$ ctest --test-dir build
```

I think we are good to go now.

**Review by kou — APPROVED** [review]
+1

Could you update the pull request description for the current status?
Our merge tool uses the pull request description as commit message.

I'll merge this once the pull request description is updated.

**Comment by kevingurney** [comment]
Yes - sure thing. I'll update the description now.

Thanks!

**Comment by kevingurney** [comment]
@kou - we updated the description of the pull request to reflect the latest status.

We also discovered a small issue during qualification. If you specified `MATLAB_BUILD_TESTS=ON`, then CMake would always use the Arrow C++ libraries built from source and ignore any custom `ARROW_HOME` or system-installed distributions of Arrow.

We just pushed a small fix for this issue in b12b6fe1398dc4ee2ebb8fb00ce54ab4bfe4012a. Let us know if you have any questions.

Thanks again!

**Comment by kou** [comment]
Thanks!
I've merged this.
